src/lib/ares_parse_caa_reply.c

```diff
@@ -36,163 +36,96 @@
 #include "ares_dns.h"
 #include "ares_data.h"
 #include "ares_private.h"
+#include <stdio.h>
 
 int
 ares_parse_caa_reply (const unsigned char *abuf, int alen,
                       struct ares_caa_reply **caa_out)
 {
-  unsigned int qdcount, ancount, i;
-  const unsigned char *aptr;
-  const unsigned char *strptr;
-  int status, rr_type, rr_class, rr_len;
-  long len;
-  char *hostname = NULL, *rr_name = NULL;
+  int status;
   struct ares_caa_reply *caa_head = NULL;
   struct ares_caa_reply *caa_last = NULL;
   struct ares_caa_reply *caa_curr;
+  cares_caa_reply *ccaa_curr = NULL;
+  cares_caa_reply *ccaa_out = NULL;
 
   /* Set *caa_out to NULL for all failure cases. */
   *caa_out = NULL;
 
-  /* Give up if abuf doesn't have room for a header. */
-  if (alen < HFIXEDSZ)
-    return ARES_EBADRESP;
-
-  /* Fetch the question and answer count from the header. */
-  qdcount = DNS_HEADER_QDCOUNT (abuf);
-  ancount = DNS_HEADER_ANCOUNT (abuf);
-  if (qdcount != 1)
-    return ARES_EBADRESP;
-  if (ancount == 0)
-    return ARES_ENODATA;
-
-  /* Expand the name from the question, and skip past the question. */
-  aptr = abuf + HFIXEDSZ;
-  status = ares_expand_name (aptr, abuf, alen, &hostname, &len);
+  status = cares_parse_caa_reply(abuf, alen, &ccaa_out);
+
   if (status != ARES_SUCCESS)
+  {
+    if (ccaa_out)
+      ares_free_data(ccaa_out);
     return status;
+  }
+
+  /* iterate through the cares_caa_reply list and
+   * create a new ares_caa_reply */
+  for(ccaa_curr = ccaa_out; ccaa_curr;
+      ccaa_curr = cares_caa_reply_get_next(ccaa_curr))
+  {
```

Comment:
For #4 on your list, I'm having trouble figuring out a way to do this loop with cares_caa_reply_get_next and others like it returning a const pointer.  The compiler throws a warning that the const value is getting ignored since I'm assigning it to a non-const variable ccaa_curr.  Is that warning okay?

If trying to make ccaa_curr const, then I would have to make it scoped to just the for loop, and make the loop an infinite loop that has an if statement inside that breaks if ccaa_curr is null.  However, then I have a problem with the very first iteration where I need to assign ccaa_curr from ccaa_out instead of cares_caa_reply_get_next.

Reply:
ccaa_curr should definitely be defined as const.  Not sure what you mean by scoped to the loop, you can just define it const where it is.  What error are you getting when you do, perhaps there's something missing in the quick review i just did..

Reply:
Okay, i tried exactly what you just said and I didn't get an error.  I think I was getting confused by constant pointer vs pointer to constant, and I was hung up thinking about how I'm mutating ccaa_curr, but really we're making the value pointed to const here, not the pointer itself. Right?

Reply:
const in this case means the data pointed to is read-only.  The pointer itself can be switched to another read-only address.